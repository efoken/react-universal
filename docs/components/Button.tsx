'use client';

import { Button as ButtonRoot } from '@react-universal/components';
import type { ButtonProps } from '@react-universal/components';
import { useCallback, useState } from 'react';
import type { MouseEvent } from 'react-native';

export const Button: React.FC<ButtonProps> = ({ onHoverIn, onHoverOut, sx, ...props }) => {
  const [hovered, setHovered] = useState(false);

  const handleHoverIn = useCallback(
    (event: MouseEvent) => {
      onHoverIn?.(event);
      setHovered(true);
    },
    [onHoverIn],
  );

  const handleHoverOut = useCallback(
    (event: MouseEvent) => {
      onHoverOut?.(event);
      setHovered(false);
    },
    [onHoverOut],
  );

  return (
    <ButtonRoot
      sx={{
        alignItems: 'center',
        backgroundColor: hovered ? 'background.muted' : 'transparent',
        borderRadius: 4,
        borderWidth: 0,
        justifyContent: 'center',
        ...sx,
      }}
      onHoverIn={handleHoverIn}
      onHoverOut={handleHoverOut}
      {...props}
    />
  );
};
