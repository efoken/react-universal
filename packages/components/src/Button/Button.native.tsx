import {
  normalizeLayoutEvent,
  normalizeResponderEvent,
  normalizeRole,
  styled,
} from '@universal-ui/core';
import { isFunction } from '@universal-ui/utils';
import { forwardRef, useState } from 'react';
import type { NativeSyntheticEvent, TargetedEvent } from 'react-native';
import { Pressable as RNPressable } from 'react-native';
import type { ButtonProps, ButtonType } from './Button.types';

const ButtonRoot = styled(RNPressable, {
  name: 'Button',
  slot: 'Root',
})(() => ({
  borderWidth: 1,
  flexDirection: 'row',
  padding: 1,
}));

export const Button = forwardRef<any, ButtonProps>(
  (
    {
      children,
      disabled = false,
      lang,
      onBlur,
      onFocus,
      onFocusVisible,
      onLayout,
      onMoveShouldSetResponder,
      onMoveShouldSetResponderCapture,
      onResponderEnd,
      onResponderGrant,
      onResponderMove,
      onResponderReject,
      onResponderRelease,
      onResponderStart,
      onResponderTerminate,
      onResponderTerminationRequest,
      onStartShouldSetResponder,
      onStartShouldSetResponderCapture,
      role,
      style,
      ...props
    },
    ref,
  ) => {
    const [focusVisible, setFocusVisible] = useState(false);
    if (disabled && focusVisible) {
      setFocusVisible(false);
    }

    const handleBlur = (event: NativeSyntheticEvent<TargetedEvent>) => {
      setFocusVisible(false);
      onBlur?.(event);
    };

    const handleFocus = (event: NativeSyntheticEvent<TargetedEvent>) => {
      setFocusVisible(true);
      onFocus?.(event);
      onFocusVisible?.(event);
    };

    const _style = ({ pressed }: { pressed: boolean }) =>
      isFunction(style)
        ? (style({ focusVisible, hovered: false, pressed }) as any)
        : (style as any);

    return (
      <ButtonRoot
        ref={ref}
        accessibilityLanguage={lang}
        disabled={disabled}
        role={normalizeRole(role) ?? 'button'}
        style={_style}
        onBlur={handleBlur}
        onFocus={handleFocus}
        onLayout={normalizeLayoutEvent(onLayout)}
        onMoveShouldSetResponder={normalizeResponderEvent(onMoveShouldSetResponder)}
        onMoveShouldSetResponderCapture={normalizeResponderEvent(onMoveShouldSetResponderCapture)}
        onResponderEnd={normalizeResponderEvent(onResponderEnd)}
        onResponderGrant={normalizeResponderEvent(onResponderGrant)}
        onResponderMove={normalizeResponderEvent(onResponderMove)}
        onResponderReject={normalizeResponderEvent(onResponderReject)}
        onResponderRelease={normalizeResponderEvent(onResponderRelease)}
        onResponderStart={normalizeResponderEvent(onResponderStart)}
        onResponderTerminate={normalizeResponderEvent(onResponderTerminate)}
        onResponderTerminationRequest={normalizeResponderEvent(onResponderTerminationRequest)}
        onStartShouldSetResponder={normalizeResponderEvent(onStartShouldSetResponder)}
        onStartShouldSetResponderCapture={normalizeResponderEvent(onStartShouldSetResponderCapture)}
        {...props}
      >
        {({ pressed }) =>
          isFunction(children) ? children({ focusVisible, hovered: false, pressed }) : children
        }
      </ButtonRoot>
    );
  },
) as unknown as ButtonType;

Button.displayName = 'Button';
