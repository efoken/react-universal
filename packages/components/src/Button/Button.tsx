'use client';

import { styled, useEventCallback, useOwnerState } from '@react-universal/core';
import { normalizeEvent, runIfFunction } from '@react-universal/utils';
import { useComposedRefs } from '@tamagui/compose-refs';
import { useRef, useState } from 'react';
import type { BlurEvent, FocusEvent } from 'react-native';
import { View } from '../View';
import type { ButtonMethods, ButtonOwnerState, ButtonProps } from './Button.types';

const ButtonRoot = styled(View, {
  name: 'Button',
  slot: 'Root',
  shouldForwardProp: (prop) => prop !== 'ownerState' && prop !== 'theme' && prop !== 'sx',
})<{
  disabled?: boolean;
  onClick?: React.MouseEventHandler<HTMLElement>;
  onMouseDown?: React.MouseEventHandler<HTMLElement>;
  onMouseEnter?: React.MouseEventHandler<HTMLElement>;
  onMouseLeave?: React.MouseEventHandler<HTMLElement>;
  ownerState: ButtonOwnerState;
  type?: 'button' | 'submit' | 'reset';
}>({
  borderWidth: 1,
  flexDirection: 'row',
  padding: 1,
  variants: {
    disabled: {
      true: {
        pointerEvents: 'box-none',
      },
      false: {
        cursor: 'pointer',
        touchAction: 'manipulation',
      },
    },
  },
});

export const Button: React.FC<ButtonProps & { ref?: React.Ref<HTMLElement & ButtonMethods> }> = ({
  as: _as,
  children,
  disabled = false,
  href,
  onBlur,
  onFocus,
  onFocusVisible,
  onHoverIn,
  onHoverOut,
  onKeyDown,
  onKeyUp,
  onPress,
  ref,
  style,
  tabIndex,
  type,
  ...props
}) => {
  const hostRef = useRef<HTMLElement>(null);

  const [hovered, setHovered] = useState<boolean>(false);
  const [pressed, setPressed] = useState<boolean>(false);

  const [focusVisible, setFocusVisible] = useState(false);
  if (disabled && focusVisible) {
    setFocusVisible(false);
  }

  const handleMouseEnter = (event: React.MouseEvent<HTMLElement>) => {
    setHovered(true);
    onHoverIn?.(normalizeEvent(event));
  };

  const handleMouseLeave = (event: React.MouseEvent<HTMLElement>) => {
    if (focusVisible) {
      event.preventDefault();
    }
    onHoverOut?.(normalizeEvent(event));
  };

  const handleBlur = useEventCallback((event: BlurEvent) => {
    if (event.target instanceof HTMLElement && !event.target.matches(':focus-visible')) {
      setFocusVisible(false);
    }
    onBlur?.(normalizeEvent(event));
  });

  const handleFocus = useEventCallback((event: FocusEvent) => {
    // Fix for https://github.com/facebook/react/issues/7769
    if (!hostRef.current) {
      // @ts-expect-error: `currentTarget` is always of type `HTMLElement`
      hostRef.current = event.currentTarget;
    }

    if (event.target instanceof HTMLElement && event.target.matches(':focus-visible')) {
      setFocusVisible(true);
      onFocusVisible?.(normalizeEvent(event));
    }

    onFocus?.(normalizeEvent(event));
  });

  const isNativeButton = () =>
    hostRef.current?.tagName === 'BUTTON' ||
    (hostRef.current?.tagName === 'A' && (hostRef.current as HTMLAnchorElement).href);

  const handleClick = (event: React.MouseEvent<HTMLElement>) => {
    if (!disabled) {
      onPress?.(normalizeEvent(event));
    }
  };

  const handleMouseDown = () => {
    if (!disabled) {
      setPressed(true);
      document.addEventListener(
        'mouseup',
        () => {
          setPressed(false);
        },
        { once: true },
      );
    }
  };

  const handleKeyDown = useEventCallback((event: React.KeyboardEvent<HTMLElement>) => {
    onKeyDown?.(event);

    if (event.defaultPrevented) {
      return;
    }

    if (event.target === event.currentTarget && !isNativeButton() && event.key === ' ') {
      event.preventDefault();
    }

    if (event.target === event.currentTarget && event.key === ' ' && !disabled) {
      setPressed(true);
    }

    // Keyboard accessibility for non interactive elements
    if (
      event.target === event.currentTarget &&
      !isNativeButton() &&
      event.key === 'Enter' &&
      !disabled
    ) {
      event.preventDefault();
      onPress?.(normalizeEvent(event));
    }
  });

  const handleKeyUp = useEventCallback((event: React.KeyboardEvent<HTMLElement>) => {
    // calling preventDefault in keyUp on a <button> will not dispatch a click event if Space is pressed
    // https://codesandbox.io/p/sandbox/button-keyup-preventdefault-dn7f0

    if (event.target === event.currentTarget) {
      setPressed(false);
    }

    onKeyUp?.(event);

    // Keyboard accessibility for non interactive elements
    if (
      event.target === event.currentTarget &&
      !isNativeButton() &&
      !disabled &&
      event.key === ' ' &&
      !event.defaultPrevented
    ) {
      onPress?.(normalizeEvent(event));
    }
  });

  const handleRef = useComposedRefs<any>(hostRef, ref);

  const ownerState = useOwnerState({
    disabled,
  });

  return (
    <ButtonRoot
      ref={handleRef}
      aria-disabled={href == null ? undefined : disabled}
      as={_as ?? (href == null ? 'button' : 'a')}
      disabled={href == null ? disabled : undefined}
      href={href}
      ownerState={ownerState}
      role="button"
      tabIndex={href == null ? tabIndex : disabled ? -1 : (tabIndex ?? 0)}
      style={runIfFunction(style, { focusVisible, hovered, pressed })}
      type={href == null ? (type ?? 'button') : undefined}
      onBlur={handleBlur}
      onClick={handleClick}
      onFocus={handleFocus}
      onKeyDown={handleKeyDown}
      onKeyUp={handleKeyUp}
      onMouseDown={handleMouseDown}
      onMouseEnter={handleMouseEnter}
      onMouseLeave={handleMouseLeave}
      {...props}
    >
      {runIfFunction(children, { focusVisible, hovered, pressed })}
    </ButtonRoot>
  );
};

Button.displayName = 'Button';
