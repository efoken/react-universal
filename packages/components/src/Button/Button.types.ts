import type { ResponderEvent, StyleProp, SxProps } from '@react-universal/core';
import type { PressableProps as RNPressableProps, ViewProps as RNViewProps } from 'react-native';
import type { ViewMethods, ViewProps, ViewStyle } from '../View';

export interface ButtonMethods extends ViewMethods {}

export interface ButtonStateCallbackType {
  readonly focusVisible: boolean;
  readonly hovered: boolean;
  readonly pressed: boolean;
}

export interface ButtonProps
  extends Omit<
      RNPressableProps,
      keyof RNViewProps | 'disabled' | 'onLongPress' | 'onPress' | 'onPressIn' | 'onPressOut'
    >,
    Omit<ViewProps, 'children' | 'style'> {
  children?: React.ReactNode | ((state: ButtonStateCallbackType) => React.ReactNode);
  disabled?: boolean;
  onFocusVisible?: RNPressableProps['onFocus'];
  onLongPress?: (event: ResponderEvent) => void;
  onPress?: (event: ResponderEvent) => void;
  onPressIn?: (event: ResponderEvent) => void;
  onPressOut?: (event: ResponderEvent) => void;
  style?: StyleProp<ViewStyle> | ((state: ButtonStateCallbackType) => StyleProp<ViewStyle>);
  /**
   * The system prop that allows defining system overrides as well as additional
   * CSS styles.
   */
  sx?: SxProps;
  type?: 'button' | 'submit' | 'reset';
}

export type ButtonOwnerState = Required<Pick<ButtonProps, 'disabled'>>;
