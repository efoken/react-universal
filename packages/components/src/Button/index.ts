export { Button } from './Button';
export type {
  ButtonMethods,
  ButtonProps,
  ButtonStateCallbackType,
} from './Button.types';
