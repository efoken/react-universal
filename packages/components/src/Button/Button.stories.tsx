/* eslint-disable react-hooks/rules-of-hooks */
import type { Meta, StoryObj } from '@storybook/react';
import { Text } from '../Text';
import { Button } from './Button';

const meta = {
  component: Button,
} satisfies Meta<typeof Button>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    children: <Text>Button</Text>,
    sx: {
      backgroundColor: '#ddd',
      borderRadius: '1',
      padding: 1,
    },
  },
};
