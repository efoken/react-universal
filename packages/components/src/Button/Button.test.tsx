import { composeStories } from '@storybook/react';
import { describe, expect, test } from 'vitest';
import * as stories from './Button.stories';

const { Default } = composeStories(stories);

describe('Button', () => {
  test('default', async () => {
    await Default.run();
    expect(document.body.firstElementChild?.firstChild).toMatchSnapshot();
  });
});
