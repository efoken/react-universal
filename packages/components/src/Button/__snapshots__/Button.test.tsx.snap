// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > default 1`] = `
<button
  class="css-r14355 css-91gi4e"
  type="button"
>
  <div
    class="css-1uf2odc"
    dir="auto"
  >
    Button
  </div>
</button>
`;
